Improve threading support in slack-libpurple.

kacf opened the pull request:
This PR adds support for three main things:

* Colorizing of threads for easier visual distinction.
* Ability to fetch thread history.
* Ability to post in threads.

This is how the thread posting works, by using the "/thread" chat command ("/th" for short):

> /th [thread-timestamp [message]]:  Post messages in threads.
> This command can be used in several ways. One can either switch focus to a thread, causing all messages to be posted to that thread, or post single messages to a thread. See these usage examples:
> - /th thread-timestamp:  Switch to thread with given timestamp.
> - /th /:  Switch back to channel.
> - /th:  Switch to the thread of the latest message. Switches to main channel if the latest reply was not threaded.
> - /th thread-timestamp message:  Post single message to the thread with the given timestamp.
> - /th / message:  Post single message to main channel.

Each commit gives more details about the implementation.

There are some things that need to be sorted out. Just to be clear: This PR is **not ready for merging**, and there are several issues I'd like to discuss first. However, it serves as a starting point, and it *does* work, so people can start testing it today.

Things to discuss:

* Is the /thread command the right way to go? The only other way I can envision is turning threads into channels, but I think this would be incredibly confusing, and I think personally I would not like this approach.

* There are some serious hacks in place to get it to work. Most notably, AFAIK there is *no* public API for getting thread replies by date if you don't already know the thread parent. This means it's impossible to get history of threads that have received replies, unless the parent is also within the same time range. The hack which is in place now is to unconditonally query the last 1000 messages (the maximum that Slack allows), even when limiting by timestamp, and do timestamp comparisons locally instead. This means that you can get thread history at least for threads whose parent is less than 1000 messages old.

* Due to normal history and thread history being different API, the history fetcher will always display normal history first, and then each thread afterwards, which means they won't be in order. This is solvable, but quite hairy because we have to remember the JSON replies and then sort them, and we don't currently have a JSON deep copy function...

* Again due to different APIs, limiting number of messages by number is not a well defined operation anymore, and if you request 1 message, you can still get hundreds of messages if that one message happened to be the parent of a busy thread. This would likely also be improved if we solved the previous point. The question is whether it is worth it, or if it's better to introduce fetching by date instead, which is often more useful.

* Posting to threads works in both channels and IMs, but in IMs it behaves a bit strangely, and this is mostly due to how Purple works (at least when using Pidgin). When switching to a thread context, and then posting, you will see both a thread-less and then a threaded message. This is because when you are not using a command, Purple posts the message immediately, without letting the plugin control how it should be displayed. Only once the server replies does it get displayed properly. I'm not sure if this is fixable, but since threads are less common in IMs, I'm not rushing to fix it.

* I've tweaked the display of thread timestamps slightly. First of all they include date, if the day is not today. Also, I have eliminated all spaces in them. This is so that they can be used with the thread command using a simple copy and paste. However, I'm not 100% sure that the time format operators I've used (%x and %X) are without whitespace in all locales, and if they're not, the "/thread" command won't work.

* If the chat background is dark instead of white (either because slack-libpurple is used in a terminal client, or if Pidgin is using a dark GTK theme), the thread colors are not well adapted. It should probably be possible to turn them off and/or enable inverted mode.

Let me know what you think!

kacf commented:
@dylex: Any interest in this, or shall I keep it in my fork? :-)

dylex commented:
What if we started with a simpler subset of this?  I like the new thread date format display, and I think it might even be okay to add the raw timestamp as a "thread id" and use that instead of typing dates.  What about starting with just one extra command to post in a thread, taking the date/thread and a message?  There could be another command to retrieve the history of a thread.

The history issue seems complicated, and it would simplify things to treat threads as entirely separate chats, but as you say, finding the threads in the first place is really messy.

kacf commented:
> What if we started with a simpler subset of this? I like the new thread date format display

Yep, makes sense. This is the most straightforward part of the pull request, so I can single that out and submit it separately.

> I think it might even be okay to add the raw timestamp as a "thread id" and use that instead of typing dates.

Yeah, ok. However, if we don't accept formatted timestamps when submitting messages, then we will have to add the raw thread id to all messages, otherwise it's not possible to start a new thread. I'm ok with that, but I think I would still vote for formatted timestamps, simply because it looks better, and we don't have to change the existing formatting. I leave it to you to decide.

> What about starting with just one extra command to post in a thread, taking the date/thread and a message?

If we assume that we will not do the "thread context" which I have implemented now, then this is a substantial deviation (the whole `_SlackThread->thread_ts` would be replaced with something more direct). I can do this, but I would rather not do it in vain, if we are anyway going to consider adding "thread context" later. Having already used this for a while, I can assure you that typing the whole command every time gets tiresome quickly in busy thread discussions! Do you think this is the wrong direction?

> There could be another command to retrieve the history of a thread.

This I can add, but I think it should be a supplement, not the only way. A very important property of this PR is that there is a way to get all history which has happened since time X. This is what happens when I log on every morning, I need to see what happened since last I was logged on, including thread replies.

What about simply replacing the current "Get history" menu item with a new item which has:

* Number of days/hours/minutes to fetch history for
* Checkbox "Fetch thread replies", which can either be on or off

This solves any ambiguity and allows you to be as detailed as you need.

kacf commented:
Regarding the thread context, this is very usage pattern sensitive, so I welcome other opinions too. I'm just listing a few people who have had an interest in the original issue https://github.com/dylex/slack-libpurple/issues/76, and in this PR: @exarkun, @fredcooke, @vn971, @blopa and @psukys.

What do you think about the ability to switch in and out of "thread focus" in a conversation? (see description of the `/th` command near the top).

fredcooke commented:
Maybe I've just used IRC in pidgin too much, but I feel like pidgin tabs have a purely linear flow and are incompatible with any kind of thread view. Slack itself shows threads in a side bar, which in pidgin terms would be another tab (not a channel, just another tab with a different purpose, I have 30 or so all the time, no big deal, irc chanserv, nickserv, actual channels, various slacks, pms, etc). 

I have no problem with some syntax to send messages in a channel and having messages display differently in the same channel would be consistent with that, but you would never be able to see the whole history of a thread without it taking away from what is going on in that channel naturally. Think in Git terms - a tree, what you are dealing with in a thread is an independent branch of conversation. 

At my previous employer we had a #devops channel and I developed a strategy of threads-only, ie each post to the channel was a summary/intro by a requester or an announcer and any discussion pertaining to that topic went in the thread under that heading post. When a task was completed it'd get a green check mark on it and you'd know it was done/complete/over/concluded. In a given day you might have had 8 threads and completed 7 of them and those 8 index posts were all still visible as a summary of "what went down today" in the channel like a table of contents. Most excellent and completely absent from half-finished gchat that I have to use in the current job :-D 

The other paradigm slack has for threads is the threads view but that's dynamic, ie, messages inserted in the  middle of the view at the end of each thread which to me doesn't work with pidgin well unless split across tabs. So I don't think having a threads tab would work well, either. 

Finally, please think about log files. I would love to see main channel log file with normal posts and thread top posts and a separate file for each thread, not the messages all mixed up in the same log file. Along those lines, consider what it means for someone to check the "send to channel to" box - not sure if that means you see it twice right now, but this is another hint at separate tabs in my eyes. 

@kacf I appreciate your efforts on this, I hope that feedback is helpful/useful and along the lines of what you were after :-)

@dylex Thanks for the work over the last year or so since I last upgraded - the invalid JSON thing killed all my accounts/sessoins yesterday or so and I upgraded to latest master and make installed and boom back up and running with no apparent bugs/bad behaviour - thank you! :-)

psukys commented:
Hey, I would side with @fredcooke  that thread should be another tab/subchannel (whatever the right terminology would be). This would as well resolve the issues  with history order fetching, and other similar issues. My only concern then would then be - getting into thread:

- should the tab be automatically created once thread is up?
- maybe a notification about thread being created could be shown in main (channel and privates) window, so that user can decide to join or not?

fredcooke commented:
@psukys That's a good point. I would say the thread should auto open and fetch history for new posts in a thread to which you're subscribed, and some sort of thread notification in the main channel for updates to ones you're not subscribed to MAYBE. Having a way to access them is important - I hadn't thought about that and it needs some careful thought and a coherent design. 

In slack to find an old thread you scroll up in the channel infinitely - not sure if this is feasible in pidgin or not, but if not, then you'd need a way to see an index of threads for a given channel with sorting available for:

* most recently updated (last post time)
* most recently created (first post time)
* maybe number of posts or other criteria

gchat sucks because no index, only whole threads popping to the bottom in most recent post time order and no option to sort by initial post time order or see only initial posts, just horrible to use. Google fanboys run the company :-D

kacf commented:
Thanks guys, it helps a lot to hear what people expect from Pidgin thread support. However, it also reveals that there is some deep rooted disagreement.

I fully agree with @fredcooke that threads do not really fit into a linear chat flow; in the way I've implemented it I consider them more as topics than threads. Everyone who has replied so far seem to agree that the best way to display them is to create separate chats for them.

I unfortunately disagree with this, and one of my main complaints about the official Slack client is that it does exactly this, hides conversations in a lot of different sub windows. I guess I just consider Slack more as a means to talk, and not as any kind of issue tracking tool, therefore I'm usually more interested in "what happened in the last hour", than "have these eight discussions been resolved or not". It's important that I can see new lines of chat by glancing at one window, not eight. I guess this is the more classical linear flow.

For huge rooms with a lot of threads, I can see that this is not the best approach, but this is not my environment.

I will leave this open for a little bit longer, in case there are more replies, but right now it looks as if my own preference for how threads should be implemented is not shared by most people. So unless I get any feedback in the next few days which supports my preference, I think I will retract this pull request, and start considering my own repository a permanent fork, where threads are implemented as inline topics, not separate chats.

kacf commented:
I can still submit the new date-based and colored thread display though, if this is of interest, since this is pretty easy to submit by itself.

fredcooke commented:
To be clear, that eight discussions with checkmarks example was just an example, mixed in with that would be numerous others from various channels, and for that I would use the threads view which I never found very good TBH. 

I think if you had only a few lightly used threads per channel then a topic approach would work okayish, but the more threads you have active in a channel at the same time, with a base "thread" IN the channel, the harder it would be to make sense of the interleaved messages even with some indicator of from whence they came. 

Question: Could this be a mode/option and mainlined with a config option to choose behaviour? I think it's inevitable that each solution would be favoured/enjoyed/used by different groups of people and that's healthy and fine IF the code base can manage it cleanly. 

Statement: I have never ever used the app, I've only ever used the slack web app and pidgin for this. pidgin has been fine, with threads the only real pain point. 

Comment: I noticed on this updated version thread replies have a little bent arrow - that's much better than the old version I had, but looks kinda messy because the names vary in length. Not sure if it's possible, but if those could be inserted LEFT of the name that'd be even cleaner/nicer. Along those lines, I could probably deal with a thread system that was one-tab based if it was clear which thread the messages came from so I could read through it all skipping lines that didn't pertain to the thread of interest and then use some command to reply into a thread reliably. But I'd still prefer separate tabs (NOT windows, that would be horrible, does slack app really do that??? lol).

psukys commented:
@kacf isn't this the behaviour of "All unread" page? I'm not entirely certain if it's just some aggregation page, that could be hooked in by API, but essentially it does exactly the thing you're describing - showing all of the changes, with ability to clean-up the queue by marking read (but essentially, it's just an event queue)

kacf commented:
> isn't this the behaviour of "All unread" page?

This page doesn't allow you to post anything AFAIK, and here you can *only* see unread messages, IOW it's not usable if you just want to revisit something somebody said a minute ago which you already saw.

> To be clear, that eight discussions with checkmarks example was just an example, mixed in with that would be numerous others from various channels, and for that I would use the threads view which I never found very good TBH.
> 
> I think if you had only a few lightly used threads per channel then a topic approach would work okayish, but the more threads you have active in a channel at the same time, with a base "thread" IN the channel, the harder it would be to make sense of the interleaved messages even with some indicator of from whence they came.

Agree, but this is no different from how a typical IRC discussion would take place, with many interleaved messages. I think what I'm trying to get at is that I don't *care* about the threads, I only care about what is being said. The threads are just an alternative to prefixing with someone's name.

I'm never interested in going back and looking at only one thread. Slack is for chatting with each other, not tracking long lived issues, so I find that anything older than a couple of hours is irrelevant.

However, I do agree that it does not scale to a large number of threads.

> But I'd still prefer separate tabs (NOT windows, that would be horrible, does slack app really do that??? lol).

Bad choice of words. I of course meant tabs, or side bars. The point I was trying to make is that it hides things from you, and you have to click in a lot of these different tabs to actually find out what was being said. There is no one window where you can both post, and see the the main chat, and see the threads, at the same time. Pidgin, OTOH, does exactly this, and this is a thing I like.

> Question: Could this be a mode/option and mainlined with a config option to choose behaviour? I think it's inevitable that each solution would be favoured/enjoyed/used by different groups of people and that's healthy and fine IF the code base can manage it cleanly.

This would be great, actually! But the question is who is going to do the work? I think the two approaches are not easily reconcilable on the code level, so there needs to be significant mindshare on both sides (and judging from this discussion I think I'm the "odd guy" when it comes to preferences).

vn971 commented:
@kacf showing everything in a continuous chat would be my preference for Pidgin as well (explanation below).

Interestingly, the reasoning is a bit different: I think pidgin is just _unable_ to have quality tab support, meaning that the tabs would be unlinked pop-ups where you can't e.g. click on a channel message to show it's thread. If that assumption on pidgin-s limitations is correct, it would be better to have a solution that's very understandable and clear in pidgin-s ecosystem, rather than trying to mimic Slack while having a different architecture.

With this reasoning kept in mind, I think there's an audience for people that would want a flat message view with the thread information only available as references.

vn971 commented:
So, should we maybe implement the basic things that everybody wants as @dylex suggested in the first comment, and when we reach something that is alternatively implemented in another PR, stop and make a preference for that? I personally see the PR-s description as a clear improvement over the current state of things

psukys commented:
> This would be great, actually! But the question is who is going to do the work? I think the two approaches are not easily reconcilable on the code level, so there needs to be significant mindshare on both sides (and judging from this discussion I think I'm the "odd guy" when it comes to preferences).

Honestly, I think the discussion we had probably could've been inside an issue, as we went a bit out of scope by discussing how thread implementation *could* look like and this PR is a great start as it lays the groundwork for threads, and subsequent PRs for whichever functionality could be presented later on.

kacf commented:
Thanks for the comments guys! Good to hear that this approach has some support. What @vn971 makes a lot of sense, we should at least be able to start with this.

So here is my suggestion:

* @dylex wanted to start with just a simple command, no focus switching. I can disable the focus switching for now.
* He also wanted to use raw timestamp values from Slack as thread identifiers, but this is not possible unless we prefix every message with the identifier, because there is no other way to get it currently. I suggest we stay with the readable timestamps.
* I add an "Fetch thread replies" checkbox in the "Get history" dialog, to prevent flooding of the chat window if thread replies are not wanted. Unchecking the checkbox then gives the old behavior.

@dylex, does this sound acceptable?

dylex commented:
This seems reasonable.  My main objection to the timestamps is the need to search for all the sub-times (.0000-.9999 or whatever), which can lead to ambiguities (multiple messages per second).  Given that you have to specify the seconds already, I wonder if we could add a ".1" to the thread message display.  That is, "1602092723.00014" could be rendered as "13:45:23.14" and accepted as an argument that way, falling back to searching for a message if it's missing.  This still makes a lot of assumptions about slack timestamp formats.

I think starting with the flat display (sort of how it is now) but adding better timestamps (that can be used directly as an argument to a /thread command) is a great start.  In retrieving history, my vote is to display only non-thread messages, but show this timestamp thread indicator on any message with a timestamp, that you can use to retrieve the thread history. Then you just need a more complicated command to start a new thread.

Regardless, I'm happy to defer to popular opinion.  My main goal is just a safe and scalable implementation.

kacf commented:
> This seems reasonable. My main objection to the timestamps is the need to search for all the sub-times (.0000-.9999 or whatever), which can lead to ambiguities (multiple messages per second). Given that you have to specify the seconds already, I wonder if we could add a ".1" to the thread message display. That is, "1602092723.00014" could be rendered as "13:45:23.14" and accepted as an argument that way, falling back to searching for a message if it's missing. This still makes a lot of assumptions about slack timestamp formats.

Unfortunately we cannot add a ".1" to the generic message display, because it is Pidgin which displays this, not the Slack plugin. So then we need to prefix all messages with an extra timestamp, and find some extra characters to differentiate chat room messages from thread messages.

I'm a bit torn on this: On one hand, it's good as you say to completely avoid all ambiguity. On the other hand, it's more noisy, and ambiguities are already handled in the code (if you try to post to a timestamp which is ambiguous, it will tell you to pick one of the raw timestamps).

But I will pick your method if I don't get any more counter arguments. At least it is very predictable!

> In retrieving history, my vote is to display only non-thread messages, but show this timestamp thread indicator on any message with a timestamp, that you can use to retrieve the thread history. Then you just need a more complicated command to start a new thread.

Yes, that makes sense, but this definitely needs to be a preference checkbox, because otherwise auto-history-fetching will not work correctly (it will silently lose thread replies).

Thinking about this a bit more holistically: I think there should just be one option in general: Enable thread display, or disable thread display. This includes both history fetching and real time chat. It would then work like this:

* When threads are disabled you won't see replies, but you will get an indicator for parent messages which have received at least one thread reply.
* The history fetching command can then override the general preference and fetch threads you are interested in.
* Thread replies which are "shared with channel" are shown in both modes.
* In the future, we could start paying attention to which threads you are subscribed to, but I think this will be in a second iteration.

dylex commented:
What if there were a command to open a thread as a separate conversation? This would avoid the need for a command to send to threads, just to "open" one. This could also fit with listing subscribed (active) threads in the conversation list at some point. (Haven't thought about exactly how this would work, but we should be able to just have a table of open threads and route messages to those conversations instead of the main one when found. Once we've settled on the basic display/thread-id infrastructure I can think about this.  Mainly just noting it for feedback and so I don't forget.)

kacf commented:
I prefer the inline display as previously mentioned, but if it were an alternative, this sounds great. Could be a preference checkbox as well.

> This could also fit with listing subscribed (active) threads in the conversation list at some point.

In a typical week I can easily participate in five new thread *per day*, so I think this would scale out of proportion very fast. A better approach might be to make them openable only by clicking on the parent message. But then it might be difficult to get back to old threads... Maybe they should not be auto added to the conversation list, for tidy-ness, but they should be available in the room list.

dylex commented:
Sorry, I didn't mean to change the default inline display, just that rather than having commands to switch and send messages on threads, there was one command to open a thread as a conversation, where you could see and reply on it, and would disable inline display for that thread.

Right, by the conversation list, I meant the room list, where I've been wanting to add something like slack's "thread" recent activity list anyway.  I agree threads should not be added to the buddy list regardless.

kacf commented:
I haven't had a lot of time to work on this lately, but I rebased this on top of the `history` branch and have implemented some of the things we agreed on. Except for simple compile fixes, everything up to and including the `Add ability to post to Slack threads.` commit is exactly the same as before.

On top of that I have implemented these things:

* The account option where you can globally turn thread display on and off.
  * This is still lacking the ability to fetch threads selectively, so disable with care. If this option is off you will not see *any* threads.
* Thread replies which are shared with channel are shown even if thread display is off.
* Thread replies are colored more brightly so it's easy to see what are thread messages and what are channel messages.
* Parent messages have received an indicator which shows whether a thread has been started on them or not.
  * An obvious weakness is that this only works properly when fetching history and the thread has already started. If you are logged in and receiving messages live, the Slack plugin can not "go back" and fix the indicator for an existing parent message. This is just how Pidgin works, and is essentially impossible to fix in a plugin.

As for future work, I plan to do these things:

* Improve history fetching so that threads can be fetched selectively even when thread display is off.
* Probably redo the history fetching dialog to input a time range instead of number of messages.

> > This seems reasonable. My main objection to the timestamps is the need to search for all the sub-times (.0000-.9999 or whatever), which can lead to ambiguities (multiple messages per second). Given that you have to specify the seconds already, I wonder if we could add a ".1" to the thread message display. That is, "1602092723.00014" could be rendered as "13:45:23.14" and accepted as an argument that way, falling back to searching for a message if it's missing. This still makes a lot of assumptions about slack timestamp formats.
> 
> Unfortunately we cannot add a ".1" to the generic message display, because it is Pidgin which displays this, not the Slack plugin. So then we need to prefix all messages with an extra timestamp, and find some extra characters to differentiate chat room messages from thread messages.
> 
> I'm a bit torn on this: On one hand, it's good as you say to completely avoid all ambiguity. On the other hand, it's more noisy, and ambiguities are already handled in the code (if you try to post to a timestamp which is ambiguous, it will tell you to pick one of the raw timestamps).
> 
> But I will pick your method if I don't get any more counter arguments. At least it is very predictable!

I have now used this feature for almost six months, and I think I can only conclude that the proposed method above isn't a good solution. Hear me out: When typing in threads, you want to type as little extra as possible. Using the mouse to copy and paste a thread identifier is disruptive, and I often end up typing it in instead, at least if it's the same day and I don't need the date. But even this length is barely at the limit of what I consider user friendly.

Now, if we insisted on no ambiguity, and added the above, it would *triple* the amount that you have to type, because you now would need both the date and the unique suffix, in addition to the time. Essentially we are making the user experience quite a lot worse just to save one API call (which is very small, we are only searching a one-second time range).

It's also worth considering that that in the six months that I have used this, an ambiguous timestamp has not occurred a single time, despite very busy rooms. That doesn't mean it can't, of course, but it means it's really rare, in practice. And the code handles it just fine if there ambiguity (it just tells you to select one of the unique ones).

In fact, I would argue that we should consider offering short forms more aggressively, for example by allowing you to specify only seconds for threads that have been opened in the last minute, only minutes and seconds for those opened in the last hour, and so on. The keyword is that typing it in should be *fast*.

Does this help in convincing you, or are you still uneasy about such ambiguity? :-) An alternative would be to merge the rest, and keep this as a unique feature in my fork, but the more we can unify, the better, I think.

kacf commented:
In addition to a couple of bug fixes:

* Add `/getthread` and a `/gt` alias, for getting history from threads.
* Improve timestamp display (less verbose)
* Add option "Display parent indicator when thread is opened". This can be enabled or disabled independently of "Display thread replies". In order to catch cases where a message initially has no replies, but then gets one, the parent message is redisplayed with a thread indicator when the first reply is posted.

Remaining work: `/getthread` doesn't work if "Display thread replies" is off, so need to find a way to override that.

I also thought of a possible method for dealing with the more experimental features: How about providing them as experimental options, that default to off? The features I'm talking about are:

* Ambiguous timestamps (e.g. `11:02:03` instead of `167189723.006000`)
* Switching thread contexts (like `cd`ing into a folder in the shell, you "cd" into a thread, and stay there until you exit it)

kacf commented:
@dylex: Alright, all the issues above are resolved. In the end I decided to remove the thread context switching from this branch, so the command is a pure thread posting command now. To summarize, this is what is added:

* New option "Display thread replies", which controls whether thread replies are displayed in the chat.
* New option "Display parent indicator when thread is opened" which displays an indicator for parent messages that have threads. This produces at most one extra message per thread, when it is first opened. This can be enabled and disabled independently of the previous option. If both are off, then you will not get any notification of thread replies at all.
* Threads are colorized.
* Unread history fetcher automatically fetches threads, if they are on.
* Normal history fetcher also does, if they are on, and one of the fetched messages is a parent.
* New command `/thread` (`/th` for short) to post to threads.
* New command `/getthread` (`/gt` for short) to fetch thread replies from individual parents. This overrides the "Display thread replies" option.

Is this acceptable?

I'm not sure how to fix the build failure in CI though. I think it is because of introducing `XOPEN_SOURCE=1` in the build, but this is required in order to get `strptime` support. The errors seem very irrelevant, they are in headers that we don't even control. Maybe `-Werror` is a bit too strict?

dylex commented:
It'd be nice if the thread timestamp formats were either configurable or followed libpurple somehow. Ideally could use `purple_str_to_time` which is a fairly flexible parser, and `purple_utf8_strftime` for formatting instead, to avoid `strptime` (which is not very portable). I'll try it out and probably have some other comments, but this definitely seems like a good set of stuff to start with.

dylex commented on slack-conversation.c:
```diff
@@ -255,15 +290,28 @@ void slack_get_history(SlackAccount *sa, SlackObject *conv, const char *since, u
 	slack_get_history(sa, conv,
 			json_get_prop_strptr(json, "last_read"),
-			json_get_prop_val(json, "unread_count", integer, -1));
+			SLACK_HISTORY_LIMIT_NUM,
```
Is there some reason not to use "unread_count" here if it's available?

kacf replied:
This is described in the original commit message for threads, but it is indeed not very discoverable. I copied that message into a comment here, to make it clearer. I encourage you to read it.

But it's good that you pointed it out, because I also realized we can optimize this: It's only needed when thread replies are turned on, so I went ahead and made this conditional.

dylex commented on slack-conversation.c:
```diff
@@ -255,15 +290,28 @@ void slack_get_history(SlackAccount *sa, SlackObject *conv, const char *since, u
 		return;
 	}
 
+	struct get_history *h = g_new(struct get_history, 1);
```
This is never freed.

kacf replied:
Good catch, fixed!

dylex commented on slack-message.c:
```diff
@@ -9,6 +10,72 @@
+
+	GString *tmp = g_string_sized_new(7);
+
+	unsigned int r = rand_r(&seed);
```
This is a slightly strange use of rand. Maybe `g_direct_hash` or even `g_str_hash` on the original ts would be a better fit.

kacf replied:
Good idea, also fixed!

dylex commented on slack-message.c:
```diff
@@ -342,18 +409,34 @@ void slack_json_to_html(GString *html, SlackAccount *sa, json_value *message, Pu
+		if (g_strcmp0(ts, thread))
+			g_string_append(html, "⤷ ");
+		else
+			g_string_append(html, "◈ ");
```
These need to be configurable. Some clients (my finch included) may not be unicode-friendly.

kacf replied:
New configuration options added! Those unicode characters are still the defaults, but I reckon that's ok, since attachments already have a unicode default.

dylex commented on slack-message.c:
```diff
@@ -366,13 +449,52 @@ void slack_json_to_html(GString *html, SlackAccount *sa, json_value *message, Pu
 			slack_attachment_to_html(html, sa, attachments->u.array.values[i]);
 }
 
+void slack_write_message(SlackAccount *sa, SlackObject *obj, const char *html, PurpleMessageFlags flags) {
```
I'm not sure why this function exists.  It seems like everywhere it's used should use `purple_conversation_write` instead since it's only for "system" messages, but maybe I'm missing something.

kacf replied:
To be perfectly honest, I don't remember why I added this, it's almost a year ago. You might be right that it can be replaced with `purple_conversation_write`.

But there is one place where it's used with `PURPLE_MESSAGE_SEND`, IOW not a system message. Is that still ok to use with `purple_conversation_write`?

dylex commented on slack-cmd.c:
```diff
@@ -138,6 +186,30 @@ void slack_cmd_register() {
+	commands = g_slist_prepend(commands, GUINT_TO_POINTER(id));
+
+	g_string_printf(thread_help, thread_fmt, "thread");
+	id = purple_cmd_register("thread", "s", PURPLE_CMD_P_PRPL, PURPLE_CMD_FLAG_IM | PURPLE_CMD_FLAG_CHAT | PURPLE_CMD_FLAG_PRPL_ONLY |
```
I think the arguments to this should be "ws" (or maybe "wS" -- been meaning to sort this out for edit), which will simplify the parsing. You also shouldn't need WRONG_ARGS.

(I'm not entirely sold on having the shorter aliases, since there's generally tab completions, but would be good to at least avoid duplication.)

kacf replied:
> I think the arguments to this should be "ws" (or maybe "wS" -- been meaning to sort this out for edit), which will simplify the parsing. You also shouldn't need WRONG_ARGS.

Yes, this is a remnant of the thread context feature, which is still being used in my local fork. I'll check out how much work it is to change this.

> (I'm not entirely sold on having the shorter aliases, since there's generally tab completions, but would be good to at least avoid duplication.)

You're right, I did not think of this. I can remove the aliases if you want, since with completion they're not really necessary.

dylex reviewed (changes requested):
Just some thoughts, but I'm also happy to do a cleanup pass myself if you want to do some of these things.

dylex commented:
The build failure isn't your fault -- github switched the environment. if you merge in history again it should pass.

kacf commented:
Thanks @dylex, I'll take a look at your comments soon! (it's a slow moving process, I don't have a lot of time to spend on this! :slightly_smiling_face:)

peci1 commented:
Outch, I liked the thread context switching functionality. I hope it will return when the time comes :)

kacf commented:
> It'd be nice if the thread timestamp formats were either configurable or followed libpurple somehow.

@dylex: What sort of configurability did you have in mind? I'm not exactly sure what you mean by "follow libpurple", but it's kind of what it tries to do. For example, take a message like this:

```
(15/04/21 11:53:37) kacf: Tada!
```

This is standard Pidgin/Purple. The Slack plugin is not involved in formatting that. But you can take that exact time/date string and make a thread out of it, except that you have to replace the space with a hyphen. So I think it might be better if it was not configurable, it would just follow the standard locale in Pidgin/Purple. This is both easier to parse and easier for users because they can copy paste verbatim strings.

(We could also consider supporting space instead of a hyphen. It makes the argument parsing a little trickier, but very nice from a usability perspective. Then it would be pure copy/paste, no change needed)

> Ideally could use `purple_str_to_time` which is a fairly flexible parser,

I did some testing of `purple_str_to_time`, and unfortunately it looks like it doesn't work for this purpose at all. It appears to ignore locales completely, so it cannot parse what Purple outputs in the chat window. And you cannot parse time-only strings, which is needed for most messages, since Pidgin only prints the time for messages that arrive the same day that they are fetched.

> and `purple_utf8_strftime` for formatting instead,

This I will use!

> to avoid `strptime` (which is not very portable).

`strptime` is in POSIX.2001, so I reckon it is quite portable. That's theory though, I don't have any practical evidence to back that up. Do you know of any practical problems with it, or is it only due to the usage of `XOPEN_SOURCE`?

EionRobb commented:
> > It'd be nice if the thread timestamp formats were either configurable or followed libpurple somehow.
> 
> @dylex: What sort of configurability did you have in mind? I'm not exactly sure what you mean by "follow libpurple", but it's kind of what it tries to do. 

I think he means to match the format that Pidgin is using.  I just checked Pidgin's code, and it uses the libpurple functions `purple_date_format_long()` and `purple_time_format()` which use `%x %X` and `%X` respectively, so I think you're all good here - although using a hypen instead of a space which is weird.

> 
> > Ideally could use `purple_str_to_time` which is a fairly flexible parser,
> 
> I did some testing of `purple_str_to_time`, and unfortunately it looks like it doesn't work for this purpose at all. 

That's true, `purple_str_to_time()` is only used to "Parses a timestamp in jabber, ISO8601, or MM/DD/YYYY format" i.e. for dealing with dates over the wire

> 
> > to avoid `strptime` (which is not very portable).
> 
> `strptime` is in POSIX.2001, so I reckon it is quite portable. That's theory though, I don't have any practical evidence to back that up. Do you know of any practical problems with it, or is it only due to the usage of `XOPEN_SOURCE`?

`strptime` isn't available in Windows/msvc.

kacf commented:
> I think he means to match the format that Pidgin is using. I just checked Pidgin's code, and it uses the libpurple functions `purple_date_format_long()` and `purple_time_format()` which use `%x %X` and `%X` respectively, so I think you're all good here - although using a hypen instead of a space which is weird.

True, the more I think about this, the more I want to fix it. However, that will force me to go back to the manual argument parsing and `WRONG_ARGS` approach that @dylex wanted gone, because it can now have either two or three arguments, depending on whether the thread is today or yesterday (with or without date). I think this justifies it though, because with this fix, copy/pasting thread times will work perfectly.

Unless you have any objections I think I will stick to the manual parsing and remove the need for the hyphen.

> `strptime` isn't available in Windows/msvc.

The second reply on [this page](https://exceptionshub.com/strptime-equivalent-on-windows.html) lists [a BSD-licensed implementation](http://cvsweb.netbsd.org/bsdweb.cgi/src/lib/libc/time/strptime.c?rev=HEAD) which might be possible to use.

However, I do not have a Windows machine to test it on, so I think I have to throw in the towel on that one. Any chance I could get some help from @EionRobb on that one? :-)

I have looked for other alternatives too, but it's hard to come up with something that matches the `%x/%X` arguments perfectly, and which is cross-platform.

dylex commented:
Ah, the display timestamps are configurable in finch (in preferences), and so I assumed they were on other clients to, but if these timestamps match the format most people will see (even if I don't), I'm happy with that.

I'm sort of okay with keeping the strptime, as long as there's some #ifdef'd  fallback case (even if it only, say, parses unix epoch time or something), so it at least compiles on windows, even if it doesn't work very well for now.

dylex commented:
I've started a pass on this on the threads branch in this repo.  I'll merge any new changes you want to make into that, and ultimately merge that into master.

chex3 commented:
Hi Guys! I am jumping on this thread as well, very excited to have improved single channel threading support in slack-libpurple, but I am using bitlbee with it.  

I have been using https://github.com/insomniacslk/irc-slack Go lang server for a couple of years now, but I just discovered this plugin in Bitlbee, which I was already using for the Slack, and Facebook gateway, so am excited to try to integrate purple-slack plugin with my other bitlbee setups I have.  

Thanks!! Stefan/Chex@Freenode

dylex commented:
I've done a full pass on the threads branch, fixed a number of issues, renamed and added some options.  Hopefully I didn't break anything in the process. If this all looks good to everyone, I'm happy to merge it in the next few days.

Obviously we can keep iterating to add more thread features. I think options both for stateful thread context and opening threads in separate chat windows would be great.

chex3 commented:
Hi, so I managed to compile this pull version of the branch, but I can't figure out how to pass this /thread command to bitlbee thru my weechat IRC client.  Is there someone working with this feature using bitlbee who can tell me how to use the /thread feature? 

Thanks, Stefan

kacf commented:
> I've done a full pass on the threads branch, fixed a number of issues, renamed and added some options. Hopefully I didn't break anything in the process. If this all looks good to everyone, I'm happy to merge it in the next few days.

@dylex: The only thing I'd like to fix (unless you disagree) before merging is the hyphen issue discussed in [this comment](https://github.com/dylex/slack-libpurple/pull/118#issuecomment-821161959). I'm trying to get this done today.

kacf commented:
> Hi, so I managed to compile this pull version of the branch, but I can't figure out how to pass this /thread command to bitlbee thru my weechat IRC client. Is there someone working with this feature using bitlbee who can tell me how to use the /thread feature?

@chex3: Sorry, I have not tried this. Do other commands work, like `/edit`?

kacf commented:
> @dylex: The only thing I'd like to fix (unless you disagree) before merging is the hyphen issue discussed in [this comment](https://github.com/dylex/slack-libpurple/pull/118#issuecomment-821161959). I'm trying to get this done today.

Oh, now I see you've already done this, nice! I will focus on testing it then.

kacf commented:
Alright, I've given this a spin, there are few issues, but overall pretty good:

* Needs [this PR](https://github.com/dylex/slack-libpurple/pull/138) to fix incorrect API call.

* ```
  commit ceb181055bdeb20c0369ac1fc05c01d2ef74add2
  Author: Dylan Simon <dylan@dylex.net>
  Date:   Sun May 2 01:17:04 2021

    Handle thread_history in the right place without breaking since
  ```
  This commit breaks history fetching via the menu. No matter what number you put in the box, it fetches 1000 messages with every thread message attached to them. It isn't really a good idea to use the 1000 message limit without also using `since`, which is only used when fetching unread messages. Ideally, the "number of messages" input box should be replaced with a "since this time" box, but for now I think reverting the commit is good enough, and we can think of more proper solutions later.

* I dug a bit more into the `%x %X` format issue, and found this:
  * In the libpurple source code, it is a translated string, and translations are allowed to swap the two arguments, but otherwise they it's always those two arguments.
  * This means that to parse them correctly, we should attempt to parse all three of `%x %X`, `%X %x` and `%X`.
  * I would like to fix this, but it's a bit trickier than I thought because of the manual parsing we have to do in the command handler.
  * I'm ok with leaving this for later, so the above is more of an FYI.

So in short, I would revert the commit I mentioned above, but otherwise I'm approving this! Finally we can get this in!

dylex commented:
For the time format, I think there are two separate concerns.  One is that translations may use different formats.  For me the larger issue is that, at least in finch and pidgin, these formats are also configurable. My solution to this was to make the thread timestamp formats also configurable, so if someone is in a "%X %x" locale, they can change the format themselves to match.

The other issue is the space/dash/undescore issue. I'm open to making the argument parsing more flexible for this. It's a little messy, but not too bad. I also want to figure out the situation with (and effect on) message formatting. But, as you say, we can leave it for the moment.

kacf commented:
Since I had some time now, I attempted to [deal with this right away](https://github.com/dylex/slack-libpurple/pull/139). With this, spaces are supported, and the configurable timestamps are also supported, so this should cover all bases I think!

The default is also changed to `%x %X`, since this is the default in Pidgin.

chex3 commented:
> > Hi, so I managed to compile this pull version of the branch, but I can't figure out how to pass this /thread command to bitlbee thru my weechat IRC client. Is there someone working with this feature using bitlbee who can tell me how to use the /thread feature?
> 
> @chex3: Sorry, I have not tried this. Do other commands work, like `/edit`?

No, /edit doesn't work either.  I feel I am missing something really basic here, we want to send commands to bitlbee, not my irc-client I am working in that is connected TO bitlbee ( and then the libpurple-slack plugin ) as a remote server.  

18:31:44               weechat =!= | Error: unknown command "/edit" (type /help for help)

I feel like this list of bitlbee commands is where I need to be sending commands thru to the plugin, do any of these ring a bell to you? 

18:37:42 @root |  * account - IM-account list maintenance
18:37:42 @root |  * channel - Channel list maintenance
18:37:42 @root |  * chat - Chatroom list maintenance
18:37:42 @root |  * add - Add a buddy to your contact list
18:37:42 @root |  * info - Request user information
18:37:42 @root |  * remove - Remove a buddy from your contact list
18:37:42 @root |  * block - Block someone
18:37:42 @root |  * allow - Unblock someone
18:37:42 @root |  * otr - Off-the-Record encryption control
18:37:42 @root |  * set - Miscellaneous settings
18:37:42 @root |  * help - BitlBee help system
18:37:42 @root |  * save - Save your account data
18:37:42 @root |  * rename - Rename (renick) a buddy
18:37:42 @root |  * yes - Accept a request
18:37:42 @root |  * no - Deny a request
18:37:42 @root |  * plugins - List all the external plugins and protocols

Thanks for any advice. 

- Stefan

SamuelNorbury commented:
I'm super happy to see this merged. Thanks everyone! :tada:

kacf commented:
Me as well, thank you @dylex!!

kacf commented:
@chex3: It looks like bitlbee maintains its own list of commands, and doesn't respect those that are added by plugins. This is more likely to be a missing feature in bitlbee.

fredcooke commented:
@dylex can you explain why this got merged to a now-non-existent branch called history and not master? Seems master was updated around the same time so I'm guessing it got there in the end, just wondering what the strategy was as a 13 year git practitioner it had me a bit confused :-D

dylex commented:
I'm not exactly sure why this PR was targeted at the history branch (probably because it was forked from it). I merged it into that branch only to make github happy, but really merged it into master, at which point both `history` and `threads` were out of date feature branches so I deleted them.

chex3 commented:
> @chex3: It looks like bitlbee maintains its own list of commands, and doesn't respect those that are added by plugins. This is more likely to be a missing feature in bitlbee.

OK, I am talking to folks in the OFTC #bitlbee channel, but people aren't being very responsive, I don't think this can get fixed soon. :( I will try to see if I can file a bug against bitlbee, anyway.  Thanks again for the help.

dakra commented:
I think this feature should also be documented in the readme for people who didn't follow this pull request.

SamuelNorbury commented:
https://github.com/dylex/slack-libpurple/pull/141

chex3 commented:
> I'm not exactly sure why this PR was targeted at the history branch (probably because it was forked from it). I merged it into that branch only to make github happy, but really merged it into master, at which point both `history` and `threads` were out of date feature branches so I deleted them.

So wait, the /history and /threads commands are no longer in the master branch of this project? I just built this pull-request again, and am running it, and from irssi/bitlbee, still cannot figure out how to get those commands to work.  

On the bright side, it seems the reconnect issues I was having with the slack API's is not an issue in irssi, its handling the reconnects properly now. 

But without threads support, I really cannot use this plugin as I want, if I want to respond to some thread, I have to go log into the slack on a browser, and reply to it there.  

-- Stefan

dylex commented:
Everything is in master currently, all the commands listed in the README.

My understanding from #144 and #142 is that bitlbee doesn't support passing slash commands through to libpurple, but that would be a question for bitlbee.  Maybe you can suggest or contribute a feature for it if they don't already. I know a number of people have asked about it.

chex3 commented:
So I talked to someone on OFTC #bitlbee, and they said they aren't taking any feature request for changes to bitlbee at all.  If someone wants to code up a PR and submit, they would be willing to take a look at it, but no one there is doing active development of bitlbee anymore, from what I understand.  

> My understanding from #144 and #142 is that bitlbee doesn't support passing slash commands through to libpurple, but that would be a question for bitlbee. Maybe you can suggest or contribute a feature for it if they don't already. I know a number of people have asked about it.
